File: Day Eleven 30-10-24/LeetCode/solution-1.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # Initialize the minimum price to the first element in the list
        minPrice = prices[0]
        # Initialize the maximum profit to 0, as no transactions have been made yet
        maxProf = 0
        
        # Iterate over each price in the list
        for price in prices:
            # If the current price is less than the minimum price encountered so far, update minPrice
            if price < minPrice:
                minPrice = price
            # Calculate the potential profit by selling at the current price after buying at minPrice
            potentialProfit = price - minPrice
            # Update maxProf if the potential profit is greater than the current maxProf
            if potentialProfit > maxProf:
                maxProf = potentialProfit
        
        # Return the maximum profit that can be achieved
        return maxProf

File: Day Eleven 30-10-24/LeetCode/solution-1.js
/**
 * Problem 1: Best Time to Buy and Sell Stock
 * 
 * You are given an array prices where prices[i] is the price of a given stock on the ith day.
 * 
 * You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.
 * 
 * Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.
 * 
 * Constraints:
 * - 1 <= prices.length <= 105
 * - 0 <= prices[i] <= 104
 */

/**
 * @param {number[]} prices - An array of stock prices where prices[i] is the price on the ith day.
 * @return {number} - The maximum profit that can be achieved from a single buy-sell transaction.
 */
const maxProfit = (prices) => {
  // Initialize the buy variable to the first day's price
  let buy = prices[0];
  // Initialize the profit variable to 0
  let p = 0;
  
  // Iterate over each price in the array
  for (let i = 0; i < prices.length; i++) {
    // If the current price is lower than the buy price, update the buy price
    if (buy > prices[i]) buy = prices[i];
    // Calculate the potential profit and update p if it's greater than the current p
    p > prices[i] - buy ? p : (p = prices[i] - buy);
  }
  
  // Return the maximum profit found
  return p;
};
